File: frontend/lib/api/metrics.ts
import { apiClient } from '@/lib/api/client';
import type { Metrics } from '@/types/metrics';

export function getMetrics(month: string) {
  return apiClient<Metrics>(`/metrics?month=${month}`);
}

// Phase 5 (2026-04-19): three-layer Flow model.
// Ref: financeapp-vault/01-Metrics/Поток.md
export interface FlowMetric {
  basic_flow: number;
  free_capital: number;          // Basic - credit_body_payments
  full_flow: number;
  cc_debt_compensator: number;   // Δ долга по КК за период
  credit_body_payments: number;  // тело обязательных платежей
  lifestyle_indicator: number | null;
  zone: 'healthy' | 'tight' | 'deficit';
  trend: number | null;
}

export interface CapitalMetric {
  capital: number;
  trend: number | null;
  trend_3m?: number | null;
  trend_6m?: number | null;
  trend_12m?: number | null;
  snapshots_count?: number;
}

export interface DTIMetric {
  dti_percent: number | null;
  zone: 'normal' | 'acceptable' | 'danger' | 'critical' | null;
  monthly_payments: number;
  regular_income: number;
}

export interface ReserveMetric {
  months: number | null;
  zone: 'critical' | 'minimum' | 'normal' | 'excellent' | null;
  available_cash: number;
  monthly_outflow: number;
}

export interface BufferStabilityMetric {
  months: number | null;
  zone: 'critical' | 'minimum' | 'normal' | 'excellent' | null;
  deposit_balance: number;
  avg_monthly_expense: number;
}

export interface MetricsSummary {
  flow: FlowMetric;
  capital: CapitalMetric;
  dti: DTIMetric;
  buffer_stability: BufferStabilityMetric;
  reserve: ReserveMetric;  // legacy compat
  fi_score: number;
}

export function getMetricsSummary() {
  return apiClient<MetricsSummary>('/metrics/summary');
}

export interface HealthRecommendation {
  metric: string;
  zone: string;
  priority: number;
  message_key: string;
  title: string;
  message: string;
}

export interface HealthSummary {
  metrics: MetricsSummary;
  fi_score: number;
  fi_zone: 'risk' | 'growth' | 'path' | 'freedom';
  weakest_metric: string;
  recommendations: HealthRecommendation[];
}

export function getHealthSummary() {
  return apiClient<HealthSummary>('/metrics/health');
}

File: app/schemas/metrics.py
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel


class FinancialIndependenceMetricResponse(BaseModel):
    percent: float
    passive_income: Decimal
    avg_expenses: Decimal
    gap: Decimal
    months_of_data: int


class SavingsRateMetricResponse(BaseModel):
    percent: float
    invested: Decimal
    total_income: Decimal


class MetricsResponse(BaseModel):
    financial_independence: FinancialIndependenceMetricResponse | None
    savings_rate: SavingsRateMetricResponse


# ── Multi-metric system (Stage 2) ───────────────────────────────────────────


class FlowMetricResponse(BaseModel):
    basic_flow: Decimal
    full_flow: Decimal
    lifestyle_indicator: float | None
    zone: str
    trend: Decimal | None


class CapitalMetricResponse(BaseModel):
    capital: Decimal
    trend: Decimal | None


class DTIMetricResponse(BaseModel):
    dti_percent: float | None
    zone: str | None
    monthly_payments: Decimal
    regular_income: Decimal


class ReserveMetricResponse(BaseModel):
    months: float | None
    zone: str | None
    available_cash: Decimal
    monthly_outflow: Decimal


class MetricsSummaryResponse(BaseModel):
    flow: FlowMetricResponse
    capital: CapitalMetricResponse
    dti: DTIMetricResponse
    reserve: ReserveMetricResponse
    fi_score: float


class HealthRecommendation(BaseModel):
    metric: str
    zone: str
    priority: int
    message_key: str
    title: str
    message: str


class HealthSummaryResponse(BaseModel):
    metrics: MetricsSummaryResponse
    fi_score: float
    fi_zone: str
    weakest_metric: str
    recommendations: list[HealthRecommendation]
